contracts/src/ModuleCRC21.sol

```diff
@@ -1,34 +1,104 @@
+        views
+    **/
+    function native_denom() public view returns (string memory) {
+        return denom;
+    }
+
+    function is_source() public view returns (bool) {
+        return isSource;
+    }
+
+
+    /**
+        Internal functions to be called by cronos module
+    **/
+    function mint_by_cronos_module(address addr, uint amount) public {
+        require(msg.sender == module_address);
+        mint(addr, amount);
+    }
+
+    function burn_by_cronos_module(address addr, uint amount) public {
+        require(msg.sender == module_address);
+        unsafe_burn(addr, amount);
+    }
+
+    function transfer_by_cronos_module(address addr, uint amount) public {
+        require(msg.sender == module_address);
+        unsafe_transfer(addr, module_address, amount);
+    }
+
+    function transfer_from_cronos_module(address addr, uint amount) public {
+        transferFrom(module_address, addr, amount);
+    }
+
+    /**
+        Evm hooks functions
+    **/
+
+    // send an "amount" of the contract token to recipient through IBC
+    function send_to_ibc(string memory recipient, uint amount) public {
+        if (isSource) {
+            transferFrom(msg.sender, module_address, amount);
+        } else {
+            unsafe_burn(msg.sender, amount);
+        }
+        emit __CronosSendToIbc(msg.sender, recipient, amount);
     }
 
     // send to another chain through gravity bridge
     function send_to_chain(address recipient, uint amount, uint bridge_fee, uint chain_id) external {
-        unsafe_burn_internal(msg.sender, add(amount, bridge_fee));
+        if (isSource) {
+            transferFrom(msg.sender, module_address, add(amount, bridge_fee));
+        } else {
+            unsafe_burn(msg.sender, add(amount, bridge_fee));
+        }
         emit __CronosSendToChain(msg.sender, recipient, amount, bridge_fee, chain_id);
     }
 
     // cancel a send to chain transaction considering if it hasnt been batched yet.
     function cancel_send_to_chain(uint256 id) external {
         emit __CronosCancelSendToChain(msg.sender, id);
     }
+
+    /**
+        Internal functions
+    **/
+
+    // unsafe_burn burn tokens without user's approval and authentication, used internally
+    function unsafe_burn(address addr, uint amount) internal {
+        // Deduct user's balance without approval
+        require(balanceOf[addr] >= amount, "ds-token-insufficient-balance");
+        balanceOf[addr] = sub(balanceOf[addr], amount);
+        totalSupply = sub(totalSupply, amount);
+        emit Burn(addr, amount);
+    }
+
+    // unsafe_transfer transfer tokens without user's approval and authentication, used internally
+    function unsafe_transfer(address src, address dst, uint amount) internal {
+        require(balanceOf[src] >= amount, "ds-token-insufficient-balance");
+        balanceOf[src] = sub(balanceOf[src], amount);
+        balanceOf[dst] = add(balanceOf[dst], amount);
+        emit Transfer(src, dst, amount);
+    }
```

Comment:
Based on the whole code, looks all good to me. Not sure if the DSToken code is the same as the current ModuleCRC20 token in `cronos-token-init-launch` repo, if so then all good.

Additional question about the `isSource` flag:
- in the current version, this flag can only be initialized once in constructor when it was deployed. It means we cannot switch contract model to the previous version of `ModuleCRC20`. Is that an expected feature? Can you provide more details?
- I'm not sure if the module_address will perform a burn operation in actual script(guess it should do to keep consistence with previous version?). And please ensure the actual burn operation() after triggered the corresponding `__CronosSendToIbc` or `__CronosSendToChain` event.
- Can you explain what this module_address actually is?

Reply:
Or I guess this module_address will transfer or burn the corresponding tokens according to the actual feature(to chain or to ibc). Like this below:
- to ibc: transfer tokens from user to module_address, then module_address burn tokens to keep consistence with previous version.
- to chain: transfer tokens from user to module_address, then module_address transfer to Gravity Bridge to the target chain(e.g. Ethereum).

Correct me if I'm wrong. Thanks!

Reply:
Or I guess you will use module_address to do different operations according to the actual use?
Based on the previous `ModuleCRC20` contract, all tokens will be burn when sending tokens:
- to ibc: 2 kinds of methods to do that, transfer to module_address then burn / transfer to module_address and directly to ibc(when getting back, directly transfer back to user, doesn't need to mint)
- to chain: 2 kinds of methods to do that, transfer to module_address then burn(mint tokens on GB side, when getting back, directly burn GB side and mint in this CRC21 side) / the same as above.

Reply:
to clarify about the source boolean.

Source define is the contract token represent an asset originated from the chain (cronos) like VVS, FERRO etc or comes from another chain (ETH, WBTC, ATOM) 

It should be defined during initialisation and should probably not change.

 
`module_address` is a special address use by the module when performing an evm operation. The module will invoke the smart contract method bypassing the orignal transaction flow and the "msg.sender" address will be set to `module_address`. It holds certains priviledge in the contract such as be able to mint or burn tokens freely or move tokens 

The reason we are not "burning/minting" tokens for `source` contract but move them instead is that we want to preserve the total supply of the tokens (calculated by total_supply()) 
For non-source tokens, burning is okay because the total supply is defined on the original contract (not in cronos)

Reply:
> Can you explain what this module_address actually is?

`module_address` is only used in:
```
require(msg.sender == module_address);
```
which guarantees it can only be called by chain native code, rather than anyone else.